libs/wire-api/src/Data/Swagger/Typed.hs

```diff
@@ -0,0 +1,144 @@
+  }
+  deriving (Functor)
+
+appendMaybe :: Maybe a -> Maybe a -> Maybe a
+appendMaybe Nothing x = x
+appendMaybe x _ = x
+
+appendNS :: S.NamedSchema -> S.NamedSchema -> S.NamedSchema
+appendNS (S.NamedSchema n1 s1) (S.NamedSchema n2 s2) =
+  S.NamedSchema (appendMaybe n1 n2) (s1 <> s2)
+
+instance Applicative NamedSchema where
+  pure _ = NamedSchema (pure (S.NamedSchema Nothing mempty)) Nothing
+  NamedSchema decl1 descr1 <*> NamedSchema decl2 descr2 =
+    NamedSchema (appendNS <$> decl1 <*> decl2) (appendMaybe descr1 descr2)
+
+instance Alternative NamedSchema where
+  empty = NamedSchema (pure (S.NamedSchema Nothing mempty)) Nothing
+  NamedSchema decl1 descr1 <|> NamedSchema decl2 descr2 =
+    NamedSchema (appendNS <$> decl1 <*> decl2) (appendMaybe descr1 descr2)
+
+runNamedSchema :: NamedSchema a -> Declare S.NamedSchema
+runNamedSchema (NamedSchema decl descr) = do
+  s <- decl
+  pure $ s & (S.schema . S.description) .~ descr
+
+instance Applicative Schema where
+  pure _ = Schema (pure mempty)
+  Schema a <*> Schema b = Schema ((<>) <$> a <*> b)
+
+instance Alternative Schema where
+  empty = Schema (pure mempty)
+  Schema a <|> Schema b = Schema ((<>) <$> a <*> b)
+
+class HasSchemaRef f where
+  schemaRef :: f a -> Declare (S.Referenced S.Schema)
+
+instance HasSchemaRef Schema where
+  schemaRef (Schema decl) = S.Inline <$> decl
+
+instance HasSchemaRef NamedSchema where
+  schemaRef tns = do
+    ns <- runNamedSchema tns
+    case ns of
+      S.NamedSchema (Just n) s -> do
+        S.declare [(n, s)]
+        pure (S.Ref (S.Reference n))
+      S.NamedSchema Nothing s ->
+        pure (S.Inline s)
+
+field' :: HasSchemaRef f => Text -> f a -> Schema a
+field' name = field name id
+
+field :: HasSchemaRef f => Text -> (S.Schema -> S.Schema) -> f a -> Schema a
+field name f s = Schema $ do
+  ref <- schemaRef s
+  let ref' = fmap f ref
+  pure $
+    mempty
+      & S.type_ ?~ S.SwaggerObject
+      & S.properties . at name ?~ ref'
+
+array :: HasSchemaRef f => f a -> Schema [a]
+array s = Schema $ do
+  ref <- schemaRef s
+  pure $
+    mempty
+      & S.type_ ?~ S.SwaggerArray
+      & S.items ?~ S.SwaggerItemsObject ref
+
+enum :: [(Value, Schema a)] -> Schema a
+enum ss =
+  Schema $
+    pure $
+      mempty
+        & S.type_ ?~ S.SwaggerString
+        & S.enum_ ?~ map fst ss
+
+named :: Text -> Schema a -> NamedSchema a
+named name (Schema s) = NamedSchema (S.NamedSchema (Just name) <$> s) Nothing
+
+unnamed :: NamedSchema a -> Schema a
+unnamed (NamedSchema s _) = Schema (view S.schema <$> s)
+
+schema :: ToTypedSchema a => NamedSchema a
+schema = toTypedSchema Proxy
+
+untypedSchema :: forall a. S.ToSchema a => NamedSchema a
+untypedSchema = NamedSchema (pure (S.toNamedSchema (Proxy :: Proxy a))) Nothing
+
+class ToTypedSchema a where
+  toTypedSchema :: Proxy a -> NamedSchema a
+
+-- to use with deriving via
+newtype TypedSchema a = TypedSchema a
+
+instance ToTypedSchema a => S.ToSchema (TypedSchema a) where
```

Comment:
Trying to explain this to myself: without `TypedSchema`, this would be an undecidable instance?  Or is there another reason for introducing it?  Can you leave a comment with the motivation?  This may be obvious to normal people, but one line of comment to help out people like me isn't too bad I hope :).

Reply:
Yes, without the `newtype` it would be an undecidable instance. Besides, an instance like
```haskell
instance ClassA a => ClassB a
```
doesn't really work too well, and can create overlapping instances in certain cases, so it's best avoided.

I guess that's why `deriving via` was introduced. So now you can do:
```haskell
data MyType = ...
  deriving ToSchema via TypedSchema MyType
```
and you get the effect you want with just one short line of boilerplate.

But yes, I will add a comment.